### PR TITLE
Update: Add snappy support on HttpContentDecoder

### LINKED ISSUE #13307: Extend decompression of HttpContentDecompressor.java
### Actual behaviour

At the moment the [HttpContentDecompressor.java](https://github.com/netty/netty/blob/cf5c467f4ead5ba5e1e36f5c4903af78bd49c35c/codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecompressor.java#LL58C9-L73) support:

- deflate
- gzip
- brotli

I wonder if we should extend it ?
I am specifically in need of supporting `snappy` 

```java
        if (GZIP.contentEqualsIgnoreCase(contentEncoding) ||
            X_GZIP.contentEqualsIgnoreCase(contentEncoding)) {
            return new EmbeddedChannel(ctx.channel().id(), ctx.channel().metadata().hasDisconnect(),
                    ctx.channel().config(), ZlibCodecFactory.newZlibDecoder(ZlibWrapper.GZIP));
        }
        if (DEFLATE.contentEqualsIgnoreCase(contentEncoding) ||
            X_DEFLATE.contentEqualsIgnoreCase(contentEncoding)) {
            final ZlibWrapper wrapper = strict ? ZlibWrapper.ZLIB : ZlibWrapper.ZLIB_OR_NONE;
            // To be strict, 'deflate' means ZLIB, but some servers were not implemented correctly.
            return new EmbeddedChannel(ctx.channel().id(), ctx.channel().metadata().hasDisconnect(),
                    ctx.channel().config(), ZlibCodecFactory.newZlibDecoder(wrapper));
        }
        if (Brotli.isAvailable() && BR.contentEqualsIgnoreCase(contentEncoding)) {
            return new EmbeddedChannel(ctx.channel().id(), ctx.channel().metadata().hasDisconnect(),
              ctx.channel().config(), new BrotliDecoder());
        }
```

I am opening this issue to talk about that.
Is it something Netty would be interested in seeing happening ?
I could be doing the PR if needed.

### Netty version

4.1

### PR DESCRIPTION
**Pull request by redcinelli**
Motivation:
I want to allow logstash http input to be able to read snappy compressed data.

Modifications:
To allow the HttpContentDecoder to support snappy it created a new control flow to test for snappy content encoding.

If it is present, we are using a specific snappy decompressor to read the data.

In order to make sure it was working I also included a e2e test on the behaviour of HttpContentDecoder.

Result:
We should be able to send snappy traffic to a server using HttpContentDocder and be ale to uncompress the traffic.

Fixes #13307

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecompressor.java [hunk 1]
```diff
@@ -72,6 +74,11 @@ protected EmbeddedChannel newContentDecoder(String contentEncoding) throws Excep
               ctx.channel().config(), new BrotliDecoder());
         }
 
+        if (SNAPPY.contentEqualsIgnoreCase(contentEncoding)){
```
Comment (hyperxpro) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecompressor.java:77]:
```suggestion
        if (SNAPPY.contentEqualsIgnoreCase(contentEncoding)) {
```

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValues.java [hunk 2]
```diff
@@ -119,6 +119,12 @@ public final class HttpHeaderValues {
      * {@code "br"}
      */
     public static final AsciiString BR = AsciiString.cached("br");
+
```
Comment (hyperxpro) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValues.java:122]:
```suggestion
 
```

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java [hunk 3]
```diff
@@ -56,6 +56,10 @@ public class HttpContentDecoderTest {
             31, -117, 8, 8, 12, 3, -74, 84, 0, 3, 50, 0, -53, 72, -51, -55, -55,
             -41, 81, 40, -49, 47, -54, 73, 1, 0, 58, 114, -85, -1, 12, 0, 0, 0
     };
+    private static final byte[] SNAPPY_HELLO_WORLD = {
+            -1, 6, 0, 0, 115, 78, 97, 80, 112, 89, 1, 16, 0, 0, 11, -66, -63,
+            -22,104, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100
```
Comment (hyperxpro) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:61]:
```suggestion
            -22, 104, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100
```

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java [hunk 4]
```diff
@@ -147,17 +151,47 @@ public void testChunkedRequestDecompression() {
...
+        assertFalse(channel.finish()); // assert that no messages are left in channel
+    }
+
+    @Test
+    public void testResponseDecompression() {
+        // baseline test: response decoder, content decompressor && request aggregator work as expected
+        HttpResponseDecoder decoder = new HttpResponseDecoder();
+        HttpContentDecoder decompressor = new HttpContentDecompressor();
+        HttpObjectAggregator aggregator = new HttpObjectAggregator(1024);
+
```
Comment (hyperxpro) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:186]:
```suggestion
```

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java [hunk 5]
```diff
@@ -147,17 +151,47 @@ public void testChunkedRequestDecompression() {
...
+
```
Comment (hyperxpro) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:187]:
```suggestion
```

**Review by hyperxpro — CHANGES_REQUESTED** [review]
Fix checkstyle.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecompressor.java [hunk 6]
```diff
@@ -72,6 +74,11 @@ protected EmbeddedChannel newContentDecoder(String contentEncoding) throws Excep
               ctx.channel().config(), new BrotliDecoder());
         }
 
+        if (SNAPPY.contentEqualsIgnoreCase(contentEncoding)) {
+            return  new EmbeddedChannel(ctx.channel().id(), ctx.channel().metadata().hasDisconnect(),
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecompressor.java:78]:
```suggestion
            return new EmbeddedChannel(ctx.channel().id(), ctx.channel().metadata().hasDisconnect(),
```

**Comment by normanmaurer** [comment]
@redcinelli can you please sign our ICLA: https://netty.io/s/icla ?

**Comment by hyperxpro** [comment]
We also need this for Http2.

**Comment by redcinelli** [comment]
> @redcinelli can you please sign our ICLA: https://netty.io/s/icla ?

Hi @normanmaurer, I believe I had done it already.
Do I get a confirmation by email ?

I have re-done it anyway just to make sure. 👌

**Pull request merged**